project_recalculate/models/project_task.py

```diff
@@ -0,0 +1,284 @@
+            # Get calendar from project
+            calendar = self.project_id.resource_calendar_id
+        elif resource and resource.calendar_id:
+            # Get calendar from assigned user
+            calendar = resource.calendar_id
+        else:
+            # Get calendar from company
+            if self.user_id.company_id:
+                # Get company from assigned user
+                company = self.user_id.company_id
+            else:
+                # If not assigned user, get company from current user
+                company = self.env.user.company_id
+            calendar = self.env['resource.calendar'].search(
+                [('company_id', '=', company.id)], limit=1)
+        return resource, calendar
+
+    def _from_days_enc(self, from_days, project_date,
+                       resource=None, calendar=None, increment=True):
+        interval = self._first_interval_of_day_get(
+            project_date, resource=resource, calendar=calendar)
+        # If project_date is holidays
+        if not interval:
+            if from_days > 0 and increment:
+                from_days += 1
+            elif from_days < 0 and not increment:
+                from_days -= 1
+            elif from_days == 0:
+                from_days = 1 if increment else -1
+        return from_days
+
+    def _from_days_dec(self, from_days, project_date,
+                       resource=None, calendar=None, increment=True):
+        if from_days == 0:
+            return 1 if increment else -1
+        interval = self._first_interval_of_day_get(
+            project_date, resource=resource, calendar=calendar)
+        # If project_date is not holidays
+        if interval:
+            if from_days > 0:
+                from_days += 1
+            elif from_days < 0:
+                from_days -= 1
+        return from_days
+
+    def _calculation_prepare(self):
+        """
+            Prepare calculation parameters:
+                - Increment=True, when task date_start is after project date
+                - Increment=False, when task date_start if before project date
+                - project_date, reference project date
+        """
+        self.ensure_one()
+        increment = self.project_id.calculation_type == 'date_begin'
+        if increment:
+            if not self.project_id.date_start:
+                raise exceptions.UserError(
+                    _('Start Date field must be defined.')
+                )
+            project_date = self.project_id.date_start
+            days = self.from_days
+        else:
+            if not self.project_id.date:
+                raise exceptions.UserError(
+                    _('End Date field must be defined.')
+                )
+            project_date = self.project_id.date
+            days = self.from_days * (-1)
+        return increment, project_date, days
+
+    def _resource_timezone(self, dt, resource=None, calendar=None):
```

Comment:
Here perhaps can use Odoo Style:
```python
    tz = 'UTC'
    if resource or calendar:
        tz = (resource or calendar).tz
    record_user_timestamp = env.user.sudo().with_context(tz=tz)
    return fields.Datetime.context_timestamp(record_user_timestamp, dt)
```

Reply:
What happens is method should not return the converted dt to the timezone of the resource or calendar. The method must return the same dt but aware of the timezone of the resource or calendar.
Example:
for
    `dt = 2019-06-23 00:00:00`
    `tz = 'Europe / Brussels'`
I want it to return
2019-06-03 00:00:00+02:00 instead of 2019-06-23 02:00:00+02:00

Reply:
I don't understand why need timezone if you like hours in UTC...

2019-06-03 00:00:00+02:00 in UTC is 2019-05-03 22:00:00+00:00